File: apps/web/lib/apollo/queries/weighing.ts
import { gql } from 'graphql-tag';

// Queries
export const GET_WEIGHING_RECORDS = gql`
  query GetWeighingRecords {
    weighingRecords {
      id
      ticketNumber
      vehicleNumber
      driverName
      vendorName
      grossWeight
      tareWeight
      netWeight
      weighingTime
      createdAt
      updatedAt
    }
  }
`;

export const GET_WEIGHING_RECORD = gql`
  query GetWeighingRecord($id: ID!) {
    weighingRecord(id: $id) {
      id
      ticketNumber
      vehicleNumber
      driverName
      vendorName
      grossWeight
      tareWeight
      netWeight
      weighingTime
      cargoType
      createdAt
      updatedAt
    }
  }
`;

// Mutations
export const CREATE_WEIGHING_RECORD = gql`
  mutation CreateWeighingRecord($input: CreateWeighingRecordInput!) {
    createWeighingRecord(input: $input) {
      id
      ticketNumber
      vehicleNumber
      driverName
      vendorName
      grossWeight
      tareWeight
      netWeight
      weighingTime
      cargoType
      createdAt
      updatedAt
    }
  }
`;

export const UPDATE_WEIGHING_RECORD = gql`
  mutation UpdateWeighingRecord($id: ID!, $input: UpdateWeighingRecordInput!) {
    updateWeighingRecord(id: $id, input: $input) {
      id
      ticketNumber
      vehicleNumber
      driverName
      vendorName
      grossWeight
      tareWeight
      netWeight
      weighingTime
      cargoType
      updatedAt
    }
  }
`;

// Subscriptions
export const WEIGHING_UPDATED_SUBSCRIPTION = gql`
  subscription WeighingUpdated {
    weighingUpdated {
      id
      ticketNumber
      vehicleNumber
      driverName
      vendorName
      grossWeight
      tareWeight
      netWeight
      weighingTime
      status
      updatedAt
    }
  }
`;

// Queue Management Queries
export const GET_WEIGHING_QUEUE = gql`
  query GetWeighingQueue {
    weighingQueue {
      id
      ticketNumber
      vehicleNumber
      driverName
      vendorName
      estimatedWeight
      arrivalTime
      status
      priority
      position
    }
  }
`;

// Reports and Analytics
export const GET_WEIGHING_SUMMARY = gql`
  query GetWeighingSummary($filters: WeighingFiltersInput) {
    weighingSummary(filters: $filters) {
      totalRecords
      totalTonnage
      averageTonnage
      completedRecords
      averageServiceTime
      efficiency
      dateRange {
        start
        end
      }
    }
  }
`;

export const GET_WEIGHING_ANALYTICS = gql`
  query GetWeighingAnalytics($period: String!, $filters: WeighingFiltersInput) {
    weighingAnalytics(period: $period, filters: $filters) {
      totalRecords
      totalTonnage
      averageTonnage
      peakHours {
        hour
        recordCount
      }
      vendorPerformance {
        vendorName
        recordCount
        averageTonnage
        averageServiceTime
      }
      trends {
        daily {
          date
          recordCount
          tonnage
        }
      }
    }
  }
`;

File: apps/golang/internal/weighing/models/weighing.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type WeighingRecord struct {
	ID            string    `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	TicketNumber  string    `gorm:"type:varchar(50);not null;uniqueIndex" json:"ticketNumber"`
	VehicleNumber string    `gorm:"type:varchar(20);not null" json:"vehicleNumber"`
	DriverName    string    `gorm:"type:varchar(100)" json:"driverName"`
	VendorName    string    `gorm:"type:varchar(100)" json:"vendorName"`
	GrossWeight   float64   `gorm:"type:decimal(10,2);not null" json:"grossWeight"`
	TareWeight    float64   `gorm:"type:decimal(10,2);not null" json:"tareWeight"`
	NetWeight     float64   `gorm:"type:decimal(10,2);not null" json:"netWeight"`
	CargoType     string    `gorm:"type:varchar(50)" json:"cargoType"`
	CompanyID     string    `gorm:"type:uuid;not null;index" json:"companyId"`
	WeighingTime  time.Time `gorm:"not null" json:"weighingTime"`
	CreatedAt     time.Time `json:"createdAt"`
	UpdatedAt     time.Time `json:"updatedAt"`
}

func (w *WeighingRecord) BeforeCreate(tx *gorm.DB) (err error) {
	if w.ID == "" {
		w.ID = uuid.New().String()
	}
	return
}
